File: backend/src/controllers/auth.controller.js
const { validationResult } = require('express-validator');
const { signup, login, getCurrentUser } = require('../services/auth.service');
const { AppError } = require('../middleware/error-handler');
const { setAuthCookie, clearAuthCookie } = require('../utils/jwt');
const { findUserById } = require('../repositories/user.repository');
const { verifyToken } = require('../utils/jwt');

async function signupHandler(req, res, next) {
  try {
    const { name, email, password } = req.body;
    const result = await signup({ name, email, password });

    const token = require('../utils/jwt').signToken({
      id: result.user.id,
      role: result.user.role,
    });

    setAuthCookie(res, token);

    return res.status(201).json({
      success: true,
      message: 'Picker account created successfully.',
      data: { user: result.user },
    });
  } catch (error) {
    return next(error);
  }
}

async function loginHandler(req, res, next) {
  try {
    const { email, password } = req.body;
    const result = await login({ email, password });

    setAuthCookie(res, result.token);

    return res.status(200).json({
      success: true,
      message: 'Login successful.',
      data: { user: result.user },
    });
  } catch (error) {
    return next(error);
  }
}

async function meHandler(req, res, next) {
  try {
    const token = req.cookies?.[process.env.COOKIE_NAME || 'picker_session'];

    if (!token) {
      throw new AppError(401, 'AUTH_REQUIRED', 'Authentication required.');
    }

    try {
      const payload = verifyToken(token);
      const user = await getCurrentUser(payload.id);

      return res.status(200).json({
        success: true,
        data: { user },
      });
    } catch (error) {
      if (error.name === 'TokenExpiredError' || error.name === 'JsonWebTokenError') {
        throw new AppError(401, 'AUTH_REQUIRED', 'Authentication required.');
      }
      throw error;
    }
  } catch (error) {
    return next(error);
  }
}

async function logoutHandler(req, res) {
  clearAuthCookie(res);
  return res.status(200).json({
    success: true,
    message: 'Logged out successfully.',
  });
}

module.exports = {
  signupHandler,
  loginHandler,
  meHandler,
  logoutHandler,
};

File: backend/controllers/auth_controller.py
from fastapi import Response
from sqlalchemy.orm import Session

from core.config import get_settings
from models.user import User
from schemas.auth import LoginRequest, SignupRequest, UserPublic
from schemas.response import SuccessResponse
from services.auth_service import AuthService

settings = get_settings()


def _set_auth_cookie(response: Response, token: str) -> None:
    response.set_cookie(
        key=settings.cookie_name,
        value=token,
        httponly=True,
        secure=settings.is_production,
        samesite="lax",
        path="/",
        max_age=settings.jwt_expire_minutes * 60,
    )


def signup(
    payload: SignupRequest, response: Response, db: Session
) -> SuccessResponse[dict]:
    service = AuthService(db)
    user, token = service.signup(payload)

    _set_auth_cookie(response, token)

    return SuccessResponse(
        message="Picker account created successfully.",
        data={"user": UserPublic.model_validate(user)},
    )


def login(payload: LoginRequest, response: Response, db: Session) -> SuccessResponse[dict]:
    service = AuthService(db)
    user, token = service.login(payload)

    _set_auth_cookie(response, token)

    return SuccessResponse(
        message="Login successful.",
        data={"user": UserPublic.model_validate(user)},
    )


def me(user: User) -> SuccessResponse[dict]:
    return SuccessResponse(data={"user": UserPublic.model_validate(user)})


def logout(response: Response) -> SuccessResponse[dict]:
    response.delete_cookie(key=settings.cookie_name, path="/")
    return SuccessResponse(message="Logged out.")
